koku/api/report/ocp/capacity/cluster_capacity.py

```diff
@@ -83,28 +85,117 @@ def capacity_annotations(self):
+                                                with cluster, node, usage_start and capacity_count fields.
+
+        Side effects:
+            - Modifies `self.count_by_cluster` to include the total capacity count for each cluster keyed by cluster.
+
+        """
+
+        unique_cluster_node_counts = {}
+        for count in node_instance_counts:
+            cluster, node, capacity_count = (count["cluster"], count["node"], count["capacity_count"])
+
+            # count each node per cluster once
+            # use max capacity count if a node has multiple values
+            if cluster not in unique_cluster_node_counts:
+                unique_cluster_node_counts[cluster] = {node: capacity_count}
+            if node not in unique_cluster_node_counts[cluster]:
+                unique_cluster_node_counts[cluster][node] = capacity_count
+            else:
+                unique_cluster_node_counts[cluster][node] = max(
+                    capacity_count, unique_cluster_node_counts[cluster].get(node, 0)
+                )
+
+        for cluster in unique_cluster_node_counts:
+            self.count_by_cluster[cluster] = sum(unique_cluster_node_counts.get(cluster).values())
+
+    def _aggregate_capacity_count_by_date(self, node_instance_counts):
+        """
+        Aggregates the maximu capacity count for each node within a cluster for each date
+        from a given set of node instances.
+
+        Args:
+            node_instance_counts (QuerySet): Django queryset containing annotated node instance counts
+                                                with cluster, node, usage_start and capacity_count fields.
+
+        Side effects:
+            - Modifies `self.count_by_date_cluster` to include the total capacity count for each cluster
+              on each date, keyed by date and then cluster.
+
+            - Also updates `self.count_by_date` to include the total aggregated capacity
+              for each date across all clusters.
+        """
+
+        daily_max_node_capacities = {}
+        for count in node_instance_counts:
+            cluster, node, usage_start, capacity_count = (
+                count["cluster"],
+                count["node"],
+                count["usage_start"],
+                count["capacity_count"],
+            )
+            # count each node per cluster per date once
+            # use max capacity count if a node has multiple values
+            usage_start_resolution = self._resolution_usage_converter(usage_start)
+
+            # composite key to access daily_max_node_capacities dict
+            cluster_date_key = (cluster, usage_start_resolution)
+            if cluster_date_key not in daily_max_node_capacities:
+                daily_max_node_capacities[cluster_date_key] = {}
+
+            daily_max_node_capacities[cluster_date_key][node] = max(
+                capacity_count, daily_max_node_capacities[cluster_date_key].get(node, 0)
+            )
```

Comment:
Same as above, some of this branching can be eliminated by using a defaultdict:
```suggestion
        daily_max_node_capacities = defaultdict(lambda: defaultdict(Decimal))
        for count in node_instance_counts:
            cluster, node, usage_start, capacity_count = (
                count["cluster"],
                count["node"],
                count["usage_start"],
                count["capacity_count"],
            )
            # count each node per cluster per date once
            # use max capacity count if a node has multiple values
            usage_start_resolution = self._resolution_usage_converter(usage_start)

            # composite key to access daily_max_node_capacities dict
            cluster_date_key = (cluster, usage_start_resolution)

            daily_max_node_capacities[cluster_date_key][node] = max(
                capacity_count, daily_max_node_capacities[cluster_date_key].get(node, 0)
            )
```